### PR TITLE
fix xgboost hook through 1.7

### PR DESCRIPTION
**Pull request by yl-to**
### Description of changes:
Tests passed
- [x] unit tests for xgboost 1.0/1.3/1.5/1.7.
- [x] higgsboson tests from 1P algo team.

#### Style and formatting:

I have run `pre-commit install && pre-commit run --all-files` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by johnbensnyder** [comment]
This looks good. Can you add some notes on what tests you've run?

**Review by johnbensnyder — APPROVED** [review]

File: smdebug/xgboost/hook.py [hunk 1]
```diff
@@ -33,8 +42,13 @@
...
 profiler_config_parser = get_profiler_config_parser(FRAMEWORK.XGBOOST)
 
+class DummyHook:
+    """class inheritance place holder for xgboost version under 1.3"""
+    def __init__(self):
+        pass
 
-class Hook(CallbackHook):
+class XGBoostBaseHook(CallbackHook):
+    """base Hook that represents a callback function in XGBoost."""
```
Comment (matherit) [inline smdebug/xgboost/hook.py:51]: Appears to be a duplicated class docstring

File: smdebug/xgboost/hook.py [hunk 2]
```diff
@@ -33,8 +42,13 @@
 
 profiler_config_parser = get_profiler_config_parser(FRAMEWORK.XGBOOST)
 
+class DummyHook:
```
[the remark this answers is not in the document]
Comment (yl-to) [reply smdebug/xgboost/hook.py:45]: I was wondering if we can merge these two child hook together like this:
```class Hook(XGBoostBaseHook, xgb.training.Callback if xgb_version >= 1.3 else DummyHook):```

[the remark this answers is not in the document]
Comment (yl-to) [reply smdebug/xgboost/hook.py:45]: Do you think above approach is viable?

[the remark this answers is not in the document]
Comment (yl-to) [reply smdebug/xgboost/hook.py:45]: But we should remove this DummyHook since its for testing.

File: smdebug/xgboost/hook.py [hunk 3]
```diff
@@ -5,24 +5,33 @@
 
 # Third Party
 import numpy as np
+import pkg_resources
```
Comment (matherit) [inline smdebug/xgboost/hook.py:8]: Should we make sure this is installed via setup.py?

File: smdebug/xgboost/hook.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matherit) [reply smdebug/xgboost/hook.py:45]: Ah, yeah I'm not sure if that would work. If you find something that does, feel free to add to PR though.

File: smdebug/xgboost/utils.py [hunk 4]
```diff
@@ -376,22 +405,38 @@ def parse_tree_model(booster, iteration, fmap=""):
...
+                    parse = fid[0].split('<')
+                    splits.append(float(parse[1]))
+                    categories.append(float("NAN"))
+                elif fid[0].find(":{") != -1:
+                    # categorical
+                    parse = fid[0].split(":")
+                    cats = parse[1][1:-1]  # strip the {}
+                    cats_split = cats.split(",")
+                    splits.append(float("NAN"))
+                    print(f"debug {cats_split}")
```
Comment (josephevans) [inline smdebug/xgboost/utils.py:425]: Should we remove this debug statement?

Comment (yl-to) [reply smdebug/xgboost/utils.py:425]: yes, will remove in next commit, thanks for catching this!

File: smdebug/xgboost/hook.py (hunk 3 above)
Comment (yl-to) [reply smdebug/xgboost/hook.py:8]: looks `pkg_resources` comes with `setuptools` package, which is used by `setup.py`
https://setuptools.pypa.io/en/latest/pkg_resources.html
Also checked xgboost container, it has pkg_resources too, I think we don't need to install it again.

**Review by matherit — APPROVED** [review]

File: smdebug/xgboost/utils.py [hunk 5]
```diff
@@ -376,22 +405,37 @@ def parse_tree_model(booster, iteration, fmap=""):
...
+                    parse = fid[0].split('<')
+                    splits.append(float(parse[1]))
+                    categories.append(float("NAN"))
+                elif fid[0].find(":{") != -1:
+                    # categorical
+                    parse = fid[0].split(":")
+                    cats = parse[1][1:-1]  # strip the {}
+                    cats_split = cats.split(",")
+                    splits.append(float("NAN"))
+                    # categories.append(cats_split if cats_split else None)
```
Comment (josephevans) [inline smdebug/xgboost/utils.py:425]: nit: remove dead code?

**Review by josephevans — APPROVED** [review]

File: smdebug/xgboost/utils.py (hunk 5 above)
Comment (yl-to) [reply smdebug/xgboost/utils.py]: removed

**Pull request merged**